_posts/2019-11-07-snowplow-javascript-tracker-2.12.0-released/index.md

```diff
@@ -0,0 +1,114 @@
+2. [Local Storage improvements](#local-storage)
+3. [Updates and bug fixes](#updates)
+4. [Upgrading](#upgrade)
+5. [Documentation and help](#doc)
+
+<!--more-->
+
+<h2 id="micro-tests">1. New test suite powered by Snowplow Micro</h2>
+
+This release introduces a new test suite into the Snowplow JavaScript Tracker which is powered by Snowplow Micro. 
+
+Snowplow Micro is a very small Snowplow Pipeline which can be initialised before running tests and then used to validate that data collection has occurred correctly. You can read more about Snowplow Micro in our earlier [blog post][snowplow-micro-post].
+
+In version 2.12.0 we have leveraged the capabilities of Snowplow Micro to build our own test suite which allows us to verify that the latest release of the Snowplow JavaScript Tracker is behaving as we expect. This gives us increased confidence in future releases of the Snowplow JavaScript Tracker.
+
+Not only that, as we recommend using Snowplow Micro to build your own test suite, this release is a an example of how to include Snowplow Micro as one part of your testing strategy. The code changes that power our new test suite can be found [here][js-tracker-tests] and will hopefully prove to be inspiration for your own Snowplow Micro powered test suite!
+
+<h2 id="local-storage">2. Local Storage improvements</h2>
+
+In this release we also focused on ensuring our use of Local Storage is robust and as reliable as our cookie storage strategy. With that in mind, we have fixed two issues that have affected use of Local Storage.
+
+1. When using the `localStorage` option of `stateStorageStrategy`, any state that was stored had no expiration time. Unlike Cookie storage which have expiry dates built in, Local Storage has no such mechanism. This effected functionality that incremented session counters, as the stored session id never expired when using Local Storage. This has now been solved by implementing expiration functionality on our local storage entries (Github [#718][718]). You can read more about the stateStorageStrategy options in our [docs][stateStorageStrategy-docs]
+
+2. We also fixed an issue in the local storage queue that the Snowplow JavaScript Tracker utilises to cache events before they are sent to the collector in a batch or to be queued if the event fails to send (for example, due to connectivity issues). Before this release, there was no upper limit on the size of the queue and this could impact the performance of the web application if it also relies on Local Storage. Most browsers only allow 5MB of Local Storage per site, so allowing control of limiting the Snowplow JavaScript Tracker use of local storage allows web developers to better decide how they wish to utilise their Local Storage limits (Github [#764][764]).
+
+By default this new limit is set to 1000 events in the queue, but it can be configured with the `maxLocalStorageQueueSize` argument when initialising the tracker.
```

Comment:
What happens to an event when the queue is full - I'm assuming they're dropped? 

Is there a tradeoff here in the event of a collector outage? If you were to set the limit to, say 100, and the collector had a blip, some events that could have ridden it out would be dropped? Might be worth noting this if it's correct.

Reply:
Yes they are dropped. This is the same as before however it now occurs earlier rather than once the whole Local Storage quota for a site is full.